// @digitalwing.co/redux-query-immutable/dist/commonjs/advanced.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.querySelectors = exports.httpMethods = exports.errorSelectors = exports.actionTypes = exports.updateResults = exports.updateEntities = exports.requestAsync = exports.mutateAsync = exports.cancelQuery = exports.queryMiddlewareAdvanced = exports.errorsReducer = exports.resultsReducer = exports.entitiesReducer = exports.queriesReducer = exports.getQueryKey = exports.connectRequest = undefined;

var _connectRequest = require('./components/connect-request');

Object.defineProperty(exports, 'connectRequest', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_connectRequest).default;
  }
});

var _queryKey = require('./lib/query-key');

Object.defineProperty(exports, 'getQueryKey', {
  enumerable: true,
  get: function get() {
    return _queryKey.getQueryKey;
  }
});

var _queries = require('./reducers/queries');

Object.defineProperty(exports, 'queriesReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_queries).default;
  }
});

var _entities = require('./reducers/entities');

Object.defineProperty(exports, 'entitiesReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_entities).default;
  }
});

var _results = require('./reducers/results');

Object.defineProperty(exports, 'resultsReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_results).default;
  }
});

var _errors = require('./reducers/errors');

Object.defineProperty(exports, 'errorsReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_errors).default;
  }
});

var _queryAdvanced = require('./middleware/query-advanced');

Object.defineProperty(exports, 'queryMiddlewareAdvanced', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_queryAdvanced).default;
  }
});

var _actions = require('./actions');

Object.defineProperty(exports, 'cancelQuery', {
  enumerable: true,
  get: function get() {
    return _actions.cancelQuery;
  }
});
Object.defineProperty(exports, 'mutateAsync', {
  enumerable: true,
  get: function get() {
    return _actions.mutateAsync;
  }
});
Object.defineProperty(exports, 'requestAsync', {
  enumerable: true,
  get: function get() {
    return _actions.requestAsync;
  }
});
Object.defineProperty(exports, 'updateEntities', {
  enumerable: true,
  get: function get() {
    return _actions.updateEntities;
  }
});
Object.defineProperty(exports, 'updateResults', {
  enumerable: true,
  get: function get() {
    return _actions.updateResults;
  }
});

var _actionTypes = require('./constants/action-types');

var actionTypes = _interopRequireWildcard(_actionTypes);

var _httpMethods = require('./constants/http-methods');

var httpMethods = _interopRequireWildcard(_httpMethods);

var _error = require('./selectors/error');

var errorSelectors = _interopRequireWildcard(_error);

var _query = require('./selectors/query');

var querySelectors = _interopRequireWildcard(_query);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.actionTypes = actionTypes;
exports.errorSelectors = errorSelectors;
exports.httpMethods = httpMethods;
exports.querySelectors = querySelectors;
