// @digitalwing.co/redux-query-immutable/dist/es/selectors/query.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.queryCount = exports.lastUpdated = exports.headers = exports.status = exports.isPending = exports.isFinished = exports.isEveryPending = exports.isEveryFinished = exports.isAnyPending = exports.isAnyFinished = undefined;

var _queryKey = require('../lib/query-key');

var TYPES = {
  SOME: 'some',
  EVERY: 'every'
};

var getQueryState = function getQueryState(queriesState, queryConfig, queryStateKey, defaultValue) {
  if (queryConfig) {
    var queryKey = (0, _queryKey.getQueryKey)(queryConfig);

    return queriesState.getIn([queryKey, queryStateKey], defaultValue);
  }
};

var getManyQueriesState = function getManyQueriesState(queriesState, queryConfig, queryStateKey, type) {
  if (queryConfig && queryConfig.keys && Array.isArray(queryConfig.keys)) {
    if (type === TYPES.EVERY) {
      return queryConfig.keys.every(function (key) {
        return queriesState.getIn([key, queryStateKey]);
      });
    }
    if (type === TYPES.SOME) {
      return queryConfig.keys.some(function (key) {
        return queriesState.getIn([key, queryStateKey]);
      });
    }
  }
};

var isAnyFinished = exports.isAnyFinished = function isAnyFinished(queriesState, queryConfig) {
  return getManyQueriesState(queriesState, queryConfig, 'isFinished', TYPES.SOME);
};

var isAnyPending = exports.isAnyPending = function isAnyPending(queriesState, queryConfig) {
  return getManyQueriesState(queriesState, queryConfig, 'isPending', TYPES.SOME);
};

var isEveryFinished = exports.isEveryFinished = function isEveryFinished(queriesState, queryConfig) {
  return getManyQueriesState(queriesState, queryConfig, 'isFinished', TYPES.EVERY);
};

var isEveryPending = exports.isEveryPending = function isEveryPending(queriesState, queryConfig) {
  return getManyQueriesState(queriesState, queryConfig, 'isPending', TYPES.EVERY);
};

var isFinished = exports.isFinished = function isFinished(queriesState, queryConfig) {
  return getQueryState(queriesState, queryConfig, 'isFinished', false);
};

var isPending = exports.isPending = function isPending(queriesState, queryConfig) {
  return getQueryState(queriesState, queryConfig, 'isPending', false);
};

var status = exports.status = function status(queriesState, queryConfig) {
  return getQueryState(queriesState, queryConfig, 'status');
};

var headers = exports.headers = function headers(queriesState, queryConfig) {
  return getQueryState(queriesState, queryConfig, 'headers');
};

var lastUpdated = exports.lastUpdated = function lastUpdated(queriesState, queryConfig) {
  return getQueryState(queriesState, queryConfig, 'lastUpdated');
};

var queryCount = exports.queryCount = function queryCount(queriesState, queryConfig) {
  return getQueryState(queriesState, queryConfig, 'queryCount');
};
